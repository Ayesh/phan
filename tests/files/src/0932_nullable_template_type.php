<?php declare(strict_types=1);

namespace App\Lib;

use TypeError;

/**
 * @template U
 */
class Test
{
    /** @var U */
    public $value;

    /**
     * @param ?U $value
     */
    public function __construct($value = null) {
        $this->value = Test::notNull($value);
    }

    /**
     * @template T
     *
     * @param ?T $subject
     * @return T
     * @throws TypeError
     */
    public static function notNull($subject)
    {
        if (null === $subject) {
            throw new TypeError(sprintf(
                '%s must not be null.',
                gettype($subject)
            ));
        }
        return $subject;
    }
}

function example(?string $s): ?int {
    $t = new Test($s);
    if (rand(0, 1)) {
        if (rand(0, 1)) {
            return (new Test(rand(0, 1) ? new \stdClass() : null))->value;
        }
        return $t->value;
    }
    return Test::notNull($s);  // should infer string, not ?string
}

