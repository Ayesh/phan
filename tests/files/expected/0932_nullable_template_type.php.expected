%s:45 PhanTypeMismatchReturn Returning new Test((rand(0, 1) ? new \stdClass() : null))->value of type \stdClass but example() is declared to return ?int
%s:47 PhanTypeMismatchReturn Returning $t->value of type string but example() is declared to return ?int
%s:49 PhanTypeMismatchReturn Returning Test::notNull($s) of type string but example() is declared to return ?int
